File: rb101_programming_foundations/rb109_general_ruby_assessment/exercises/101-109_review/easy/alphabetical_numbers.rb
# Source: https://launchschool.com/exercises/c688f4e5

# Write a method that takes an Array of Integers between 0 and 19, and returns
# an Array of those Integers sorted based on the English words for each number:

def alphabetical_number_sort(num_strs, nums)
  hsh = [nums, num_strs].transpose.to_h
  nums.sort_by { |num| hsh[num] }
end

strs = %w(zero one two three four five six seven eight nine ten eleven
          twelve thirteen fourteen fifteen sixteen seventeen eighteen
          nineteen)
nums = (0..19).to_a

p alphabetical_number_sort(strs, nums) == [
  8, 18, 11, 15, 5, 4, 14, 9, 19, 1, 7, 17,
  6, 16, 10, 13, 3, 12, 2, 0
]

File: exercises/javascript_exercises/210-javascript_smallproblems/list_processing/alphabetical_numbers.js
// Write a function that takes an array of integers between 0 and 19, and returns
// an array of those integers sorted based on the English word for each number:
// zero, one, two, three, four, five, six, seven, eight, nine, ten, eleven, twelve,
// thirteen, fourteen, fifteen, sixteen, seventeen, eighteen, nineteen

const ORDER = ("zero, one, two, three, four, five, six, seven, " +
               "eight, nine, ten, eleven, twelve, thirteen, fourteen, " +
               "fifteen, sixteen, seventeen, eighteen, nineteen").split(", ")

function alphabeticNumberSort(numbers) {
  return numbers.sort(sortNumbers);
}

const sortNumbers = function(num1, num2) {
    if (ORDER[num1] > ORDER[num2]) {
    return 1;
  } else if (ORDER[num1] < ORDER[num2]) {
    return -1;
  } else {
    return 0;
  }
}

let numbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19];
console.log(alphabeticNumberSort(numbers));